src/test/java/com/google/devtools/build/lib/skyframe/TrackSourceDirectoriesIntegrationTest.java

```diff
@@ -0,0 +1,93 @@
+package com.google.devtools.build.lib.skyframe;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertContainsEvent;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertDoesNotContainEvent;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.buildtool.util.BuildIntegrationTestCase;
+import com.google.devtools.build.lib.events.EventKind;
+import com.google.devtools.build.lib.events.util.EventCollectionApparatus;
+import com.google.devtools.build.lib.vfs.Path;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link TrackSourceDirectoriesFlag}. */
+@RunWith(JUnit4.class)
+public final class TrackSourceDirectoriesIntegrationTest extends BuildIntegrationTestCase {
+
+  @Override
+  protected EventCollectionApparatus createEvents() {
+    return new EventCollectionApparatus(ImmutableSet.of(EventKind.FINISH));
+  }
+
+  @Before
+  public void setupAndBuildTestTarget() throws Exception {
+    // Enabled via jvm_flags.
+    assertThat(TrackSourceDirectoriesFlag.trackSourceDirectories()).isTrue();
+
+    getWorkspace().getRelative("pkg/dir/empty_dir").createDirectoryAndParents();
+    write("pkg/dir/file", "foo");
+    write("pkg/BUILD",
+        "genrule(",
+        "    name = 'a',",
+        "    srcs = ['dir'],",
+        "    outs = ['out'],",
+        "    cmd = 'touch $@',",
+        ")");
+
+    // Build the target once - all tests below modify the source directory and assert that this
+    // causes the target to be rebuilt.
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void noChanges_targetDoesNotRebuild() throws Exception {
+    assertTargetDoesNotRebuild();
+  }
+
+  @Test
+  public void fileCreatedInEmptyDir_targetDoesRebuild() throws Exception {
+    write("pkg/dir/empty_dir/file");
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void fileContentChanged_targetDoesRebuild() throws Exception {
```

Comment:
Something to consider, it's about my personal preference, so please only apply if you actually like that better.

What I would consider here is having really 2 test cases, no change and change. The latter would be parameterized by the kind of change:

```
private enum Change {
  CREATE_EMPTY_DIRECTORY {
    @Override
    void apply(Path sourceDirectory) throws IOException {
      sourceDirectory.getRelative("empty_dir/nested_empty_dir").createDirectory();
    }
  }, /* ... */
  
  abstract void apply(Path sourceDirectory) throws IOException;
}

@Test
public void build_changedSourceDirectory_rebuildsTarget(@TestParameter Change change) {
  getWorkspace().getRelative("pkg/dir/empty_dir").createDirectoryAndParents();
  write("pkg/dir/file", "foo");
  write("pkg/BUILD",
      "genrule(",
      "    name = 'a',",
      "    srcs = ['dir'],",
      "    outs = ['out'],",
      "    cmd = 'touch $@',",
      ")");
   // Initial build
   buildTarget("//foo:a");
   assertContainsEvent(events.collector(), "Executing genrule //pkg:a");
   events.collector().clear();
 
   change.apply(getWorkspace().getRelative("pkg/dir"));
   buildTarget("//foo:a");

   assertContainsEvent(events.collector(), "Executing genrule //pkg:a");
}
```

For the unchanged one, I would just either just copy-paste the setup part to it or include an assert method on the `@TestParameter`, likely the former would actually work better. The upside here is that the test only uses common and known helpers and concise test case demonstrates all that is going on here.

Reply:
Done.